### ISSUE TITLE
upgrade from 0.16.2 to 0.17.2 destroys skimage.external

### ISSUE THREAD

**Issue by DenisPolygalov** [scikit-image/scikit-image#5140]
## Description

I'm using conda environment on Windows 10 with only default channel enabled.
No mix of pip-derived and conda-derived packages or manually compiled packages present in the environment.
Before upgrading from 0.16.2 to 0.17.2 everything was working correctly.
After upgrade by using "conda update --all" skimage.external dissappear.
The folder "external" is absent on the disk (in site-packages/skimage folder of the virtual environment in question).

## Way to reproduce
```python
conda create -n env1
conda activate env1
conda install scikit-image
python -c "from skimage.external import tifffile"
```
The code above cause ModuleNotFounError: No module named `skimage.external'

## Version information
```python
# Paste the output of the following python commands
from __future__ import print_function
import sys; print(sys.version)
import platform; print(platform.platform())
import skimage; print("scikit-image version: {}".format(skimage.__version__))
import numpy; print("numpy version: {}".format(numpy.__version__))
```

```python
# output of the code above:
3.8.5 (default, Sep  3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)]
Windows-10-10.0.18362-SP0
scikit-image version: 0.17.2
numpy version: 1.19.2
```

**Comment by jni** [comment]
Thanks for reporting @DenisPolygalov. It was certainly an oversight on our part that we never considered `skimage.external` part of the public API, so we simply removed it. I presume you are after tifffile, though, and if so, the solution is to replace `from skimage.external import tifffile` with simply `import tifffile` — and depend on `tifffile` itself. Could you let us know whether that fixes the problem for you?

**Comment by DenisPolygalov** [comment]
I see... I didn't know that `skimage.external` is not a part of public API...
Replacing `from skimage.external import tifffile` with just `import tifffile` will solve problems for the case of running my Python code in a 64-bit conda environment **but break the code for 32-bit conda environment** because in the 32-bit conda there is still 0.16.2 version of scikit-image and it is the most recent. The `import tifffile` doesn't work in scikit-image 0.16.2. Can you update scikit-image to 0.17.2 in 32-bit conda package repository too?

**Comment by hmaarrfk** [comment]
Are you using anaconda repos or conda-forge?

The requests to update the package should go to those respective places.

Windows 32bit is by and large getting phased out of community support.

**Comment by DenisPolygalov** [comment]
> Are you using anaconda repos or conda-forge?

I'm using only anaconda repos in principle. Allowing conda-forge breaks more things than fixes IMHO.

> The requests to update the package should go to those respective places.

I see. So I guess this should be closed.

> Windows 32bit is by and large getting phased out of community support.

First of all let me emphasize that "32-bit version of Windows" and "running 32-bit version of Python packages on 64-bit Windows" are two slightly different things and explain my situation. My code heavily depend on two things: tiff-file support from scikit-image and QtMultimedia support from PyQt5. The QtMultimedia is **broken for years in 64-bit** PyQt Anaconda package: https://github.com/ContinuumIO/anaconda-issues/issues/10250 So if Anaconda repo managers will not update 32-bit version of scikit-image from 0.16.2 to 0.17.2 as you suggested above my code will be efficiently rendered useless because of interlock of these two bugs or I will need to use nasty tricks in my code such as different behavior depending on the platform (32/64) used.

**Comment by jni** [comment]
@DenisPolygalov Sorry, I'm confused... If there is no 32-bit skimage 0.17 and you *need* 32-bit support, then why was the upgrade attempted by conda? At any rate, have you tried PyQt5 (pyqt) or PySide2 (pyside) on conda-forge instead of Anaconda (defaults)? I think if you can somehow get your application working on 64-bit you will have a much smoother experience going forward. We don't control the Anaconda channels and iirc we also don't control which binaries get built on conda-forge — we just follow what the community is doing.

Another option might be to move some of your package management to pip instead of conda: I just saw that we are still building 32-bit Win wheels as is tifffile (pure Python now) and [imagecodecs](https://pypi.org/project/imagecodecs/#files)

**Comment by DenisPolygalov** [comment]
> If there is no 32-bit skimage 0.17 and you need 32-bit support, then why was the upgrade attempted by conda?

My approach is to use only single package manager (conda) only default package distribution channel and periodically test my code on both 32/64 bit environments. This time it just happen to be a changes in scikit-image API that was a big surprise to me and it turns out that versions are different.

> At any rate, have you tried PyQt5 (pyqt) or PySide2 (pyside) on conda-forge instead of Anaconda (defaults)?

I will try that only if I have no other choice. To my experience using non-default channels while may seem to be a solution for immediate problem will cause disaster in a long term.

> I think if you can somehow get your application working on 64-bit you will have a much smoother experience going forward.

I agree, but unfortunately absence of QtMultimedia on default 64bit conda channel is quite big obstacle for me.

> We don't control the Anaconda channels and iirc we also don't control which binaries get built on conda-forge

Oh, this is completely OK, I understand. I mean I don't have anything to say to scikit-image developers and the next step I guess is to try to talk to people who are updating packages for conda... The only thing I didn't quite get is the fact that `skimage.external.tifffile` is the part of the scikit-image documentation all the way up to 0.16.x : https://scikit-image.org/docs/0.16.x/api/skimage.external.tifffile.html  and therefore is (at least in my head) the part of official public APIs. This contradict with the statement by @jni above: 

> we never considered skimage.external part of the public API

or maybe the real question here is why doing only `import tifffile` does not work in any version before 0.17.x. ?
Regards,
Denis.

**Comment by jni** [comment]
> To my experience using non-default channels while may seem to be a solution for immediate problem will cause disaster in a long term.

*Mixing* channels will often cause issues, but using conda-forge exclusively is ok in general — I have not had issues with that so far...

> The only thing I didn't quite get is the fact that skimage.external.tifffile is the part of the scikit-image documentation all the way up to 0.16.x : scikit-image.org/docs/0.16.x/api/skimage.external.tifffile.html and therefore is (at least in my head) the part of official public APIs. This contradict with the statement by @jni above:

There is no contradiction, apologies for my unclear phrasing: you are absolutely right that it *was* part of our public API. What I meant to say is that people in the core team, myself included, did not *think about it* in that way. So when we removed it we did so without a deprecation cycle. (Note that the documentation is automatically generated — I don't think anyone made the conscious decision to document skimage.external on our API docs.) My apologies! That was 100% a mistake.

I am just trying to figure out how best to move forward. Since we have just released scikit-image 0.18.0, I expect that that will be the next version included in Anaconda, and it is *probably* (but not 100% sure here) not worth the effort to update 0.17.x, since there is a good chance you won't see the benefits of that anyway. See below for a proposal:

> the real question here is why doing only import tifffile does not work in any version before 0.17.x.

tifffile is an external package, not managed by us. I just saw that it is available on the defaults channel:

```
 $ conda search -c defaults tifffile
Loading channels: done
# Name                       Version           Build  Channel
tifffile                      0.15.1  py36h5ca1d4c_2  pkgs/main
tifffile                      0.15.1  py36h776bbcc_2  pkgs/main
tifffile                      0.15.1  py37h5ca1d4c_2  pkgs/main
tifffile                      0.15.1  py37h776bbcc_2  pkgs/main
tifffile                      0.15.1  py38h5ca1d4c_2  pkgs/main
tifffile                      0.15.1  py38h776bbcc_2  pkgs/main
tifffile                   2020.9.22  py36h0cf3a3e_2  pkgs/main
tifffile                   2020.9.22  py37h0cf3a3e_2  pkgs/main
tifffile                   2020.9.22  py38h0cf3a3e_2  pkgs/main
tifffile                   2020.10.1  py36h0cf3a3e_2  pkgs/main
tifffile                   2020.10.1  py37h0cf3a3e_2  pkgs/main
tifffile                   2020.10.1  py38h0cf3a3e_2  pkgs/main
tifffile                   2020.10.1            py_0  pkgs/main
tifffile                  2020.11.26    pyhd3eb1b0_0  pkgs/main
tifffile                   2020.12.4    pyhd3eb1b0_0  pkgs/main
tifffile                   2020.12.8    pyhd3eb1b0_0  pkgs/main
```

The reason that it is not working with a previous version is that in 0.17 we stopped maintaining our own copy in external and *instead* we added a dependency to the actual, not-maintained-by-us package. In 0.16, we used it if it was installed separately, but if not then we used our local copy (the one in .external). We did not depend on it, so `conda install scikit-image=0.16` installs scikit-image but not tifffile, while `conda install scikit-image=0.17` installs both. `conda install scikit-image=0.16 tifffile` should again install both.

So, I *think* all your problems will be solved if you add a dependency to tifffile for your package. You should then be able to `import tifffile` both with skimage 0.16 and 0.17.

Could you tell me whether that works for you?

**Comment by DenisPolygalov** [comment]
> So, I think all your problems will be solved if you add a dependency to tifffile for your package. You should then be able to import tifffile both with skimage 0.16 and 0.17.
> Could you tell me whether that works for you?


I just tested the external tifffile package from default channel. It works for me, thanks.
